crates/starknet-types-core/src/hash/poseidon.rs

```diff
@@ -0,0 +1,58 @@
+use crate::felt::Felt;
+use lambdaworks_crypto::hash::poseidon::{
+    starknet::PoseidonCairoStark252, Poseidon as PoseidonLambdaworks,
+};
+use lambdaworks_math::field::{
+    element::FieldElement, fields::fft_friendly::stark_252_prime_field::Stark252PrimeField,
+};
+
+use super::traits::StarkHash;
+
+pub struct Poseidon;
+
+impl StarkHash for Poseidon {
+    /// Computes the Poseidon hash of two Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Hashing/hash-functions/#poseidon_hash.>
+    fn hash(felt_0: &Felt, felt_1: &Felt) -> Felt {
+        Felt(PoseidonCairoStark252::hash(&felt_0.0, &felt_1.0))
+    }
+    /// Computes the Poseidon hash of an array of Felts, as defined
+    /// in <https://docs.starknet.io/documentation/architecture_and_concepts/Cryptography/hash-functions/#poseidon_array_hash.>
+    fn hash_array(felts: &[Felt]) -> Felt {
+        Felt(PoseidonCairoStark252::hash_many(unsafe {
+            core::mem::transmute::<&[Felt], &[FieldElement<Stark252PrimeField>]>(felts)
+        }))
+    }
```

Comment:
Don't use transmute, too much weird behaviours (alignment and so on).
You can take inspiration from this snippet of madara:
```rust
fn convert_calldata(data: &[Felt252Wrapper]) -> &[FieldElement] {
    // Non-copy but less dangerous than transmute
    // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
    unsafe { core::slice::from_raw_parts(data.as_ptr() as *const FieldElement, data.len()) }
}
```